### PR TITLE
Refactor tests by theme

### LINKED ISSUE #29: Arrange tests by topic matter rather than by comparison function
tests/examples.py is a file that contains several lisp-formatted test cases. This is a legacy file from the early days of Python-WikipediaBase, where it served as a functional specification.

Tests in tests/examples.py contain all sorts of test cases that are not well organized. The test output is not specific, and many test cases aren't well-commented.

Only test cases for Resolvers should go in tests/test_resolvers.py (`get` and `get-attribute` tests). Other examples, like `sort-symbols` or `get-classes`, should go in their respective files.

Each resolver should get its own test suite, e.g. `class TestPersonResolver(unittest.TestCase)`.

Method names should be descriptive, so if a test fails we know what's going on. For example, `test_currently_alive` is better than `test_obama`.

### PR DESCRIPTION
**Pull request by sherryy**
Refactor tests from examples.py to test_resolvers.py and test_classifiers.py

Fixes #29

### CODE REVIEW

File: tests/test_resolvers.py [hunk 1]
```diff
@@ -49,12 +43,420 @@ def setUp(self):
...
+                                 r'influence of mothers in society')
+        # tests that URLs are fully qualified
+        self.assertRegexpMatches('(get "wikibase-term" "Sacagawea" (:CODE "SHORT-ARTICLE"))',
+                                 r'<a href=\\"https://en.wikipedia.org/wiki/Lewis_and_Clark_Expedition')
+
+    # TODO : debug why the examples below fail spuriously and uncomment them
+    # the tests occasionally fail due to a suspected race condition with lxml
+    # See issue #35 for more context
+    @unittest.expectedFailure
+    def test_regex_match_from_infobox(self):
```
Comment (chsamlee) [inline tests/test_resolvers.py]: Should move all InfoboxResolvers test into a separate class.

File: tests/test_resolvers.py [hunk 2]
```diff
@@ -49,12 +43,420 @@ def setUp(self):
         self.kb = KnowledgeBase()
         self.term_resolver = resolvers.TermResolver()
 
+    def tearDown(self):
+        pass
+
+    def get_query_result(self, query):
+        return self.fe.eval(query)
+
+    # def _ans_match(self, lst, just=None):
```
Comment (chsamlee) [inline tests/test_resolvers.py]: Remove the code if we are not using it anymore.

File: tests/test_resolvers.py [hunk 3]
```diff
@@ -49,12 +43,420 @@ def setUp(self):
...
+        self.assertNotRegexpMatches('(get "wikibase-term" "North America" (:code "SHORT-ARTICLE"))',
+                                    r'Uploads')
+
     def test_word_count(self):
         self.assertEqual(len(self.term_resolver._words(ARTICLE_BODY)), 100)
 
     def test_url(self):
         url = self.term_resolver.resolve('wikibase-term', 'Barack Obama', {'code': 'url'})
-        self.assertEquals('https://en.wikipedia.org/wiki/Barack_Obama', url.val)
+        self.assertEqual('https://en.wikipedia.org/wiki/Barack_Obama', url.val)
```
Comment (chsamlee) [inline tests/test_resolvers.py]: Why the change from assertEquals to assertEqual? Are there any differences?

File: tests/test_resolvers.py [hunk 4]
```diff
@@ -49,12 +43,420 @@ def setUp(self):
...
+        # infobox defined outside the article, in its own template
+        self.assertEquals('(get "wikipedia-military-conflict" "World War I" (:code "DATE"))',
+                          ('((:yyyymmdd 19140728))'))
+
+        self.assertEquals('(get "wikipedia-mountain" "Mount Everest" (:code "ELEVATION-M"))',
+                          '((:html "8848"))')
+        self.assertEquals('(get "wikipedia-president" "Bill Clinton" (:code "SUCCESSOR"))',
+                          '((:html "George W. Bush"))')
+
+    def test_special_attribute_from_article(self):
```
Comment (chsamlee) [inline tests/test_resolvers.py]: Maybe break down this test case further more by attribute (number, proper, ...)

**Comment by chsamlee** [comment]
LGTM!

File: tests/test_classifiers.py [hunk 5]
```diff
@@ -132,5 +133,54 @@ def test_sections(self):
...
+        return fe.eval(query)
+
+    def test_get_class(self):
+        # =====================
+        # tests for get-classes
+        # =====================
+
+        # Would be clearer as something like:
+        #   (assert-member 'wikipedia-term' '(get-classes \"Bill Clinton\")')
+        self.assertRegexpMatches(self.get_query_result('(get-classes \"Bill Clinton\")'),
```
Comment (alvaromorales) [inline tests/test_classifiers.py]: No need to escape double quotes (`\"`). Because you're defining the query string using single quotes, you can just write:

```python
self.assertRegexpMatches(self.get_query_result('(get-classes "Bill Clinton")'), ...)
```

Please update this for all tests below.

File: tests/test_classifiers.py [hunk 6]
```diff
@@ -132,5 +133,54 @@ def test_sections(self):
...
+                                 r'wikibase-sections',
+                                 'All Wikipedia symbols have calculated class wikibase-sections'),
+        self.assertRegexpMatches(self.get_query_result('(get-classes \"Bill Clinton\")'),
+                                 r'wikipedia-president'),
+        self.assertRegexpMatches(self.get_query_result('(get-classes \"Ada (programming language)\")'),
+                                 r'wikipedia-programming-language'),
+
+        # infobox defined outside the article, in its own template
+        self.assertRegexpMatches(self.get_query_result('(get-classes \"World War I\")'),
```
Comment (alvaromorales) [inline tests/test_classifiers.py]: make this test into a separate method: `test_external_infobox`

File: tests/test_classifiers.py [hunk 7]
```diff
@@ -132,5 +133,54 @@ def test_sections(self):
...
+                                 r'wikipedia-military-conflict'),
+
+        # recognize taxoboxes
+        self.assertRegexpMatches(self.get_query_result('(get-classes \"Ulva anandii\")'),
```
Comment (alvaromorales) [inline tests/test_classifiers.py]: make this test into a separate method: `test_recognize_taxoboxes`

File: tests/test_classifiers.py [hunk 8]
```diff
@@ -132,5 +133,54 @@ def test_sections(self):
...
+                                 r'wikipedia-taxobox'),
+
+        # ===============================================================
+        # Tests that infobox attributes that aren't dates but may contain
+        # dates aren't returned in yyyymmdd format
```
Comment (alvaromorales) [inline tests/test_classifiers.py]: ^ this comment doesn't apply to the tests below. Remove it?

File: tests/test_resolvers.py [hunk 9]
```diff
@@ -37,12 +36,6 @@
 DISABLE_ALL_EXCEPT_JUST = False
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: remove this variable, it's unused.

File: tests/test_resolvers.py [hunk 10]
```diff
@@ -51,40 +44,239 @@ def setUp(self):
...
 
-    def test_get_not(self):
-        for ans, rx, msg in self._ans_match(WIKI_EXAMPLES_NOT):
-            self.assertNotEqual(ans, rx, msg=msg)
+class TestPersonResolvers(TestResolvers):
 
-    def test_get_rx(self):
-        for ans, rx, msg in self._ans_match(WIKI_EXAMPLES_RX):
-            self.assertRegexpMatches(ans, rx, msg=msg)
+    def setUp(self):
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: No need to subclass a method if all you do is call the `super` method. Remove `setUp` and `tearDown` from `TestTermResolvers`

File: tests/test_resolvers.py [hunk 11]
```diff
@@ -51,40 +44,239 @@ def setUp(self):
...
+                          '((:calculated :masculine))')
+        self.assertEquals('(get "wikibase-person" "Sacagawea" (:code "GENDER"))',
+                          '((:calculated :feminine))')
+        self.assertEquals('(get "wikibase-person" "Mary Shakespeare" (:calculated "GENDER"))',
+                          '((:calculated :feminine))')
+
+        # TODO : find an example of a person who uses gender-neutral pronouns
+
+        # we incorrectly report :neuter for male Siamese twins
+        self.assertEquals('(get "wikibase-person" "Chang and Eng Bunker" (:calculated "GENDER"))',
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: Move this into a separate test

```python
@unittest.expectedFailure
def test_gender_of_twins(self):
    self.assertEquals('(get "wikibase-person" "Chang and Eng Bunker" (:calculated "GENDER"))'
                      '((:calculated :masculine))')

File: tests/test_resolvers.py [hunk 12]
```diff
@@ -51,40 +44,239 @@ def setUp(self):
...
+        self.assertEquals('(get "wikibase-person" "Manno Wolf-Ferrari" (:code "BIRTH-DATE"))',
+                          '((:yyyymmdd 19110000))')
+        self.assertEquals('(get "wikibase-person" "Manno Wolf-Ferrari" (:code "DEATH-DATE"))',
+                          '((:yyyymmdd 19940000))')
+        self.assertEquals('(get "wikibase-person" "Freidun Aghalyan" (:code "BIRTH-DATE"))',
+                          '((:yyyymmdd 18761120))')
+        self.assertEquals('(get "wikibase-person" "Freidun Aghalyan" (:code "DEATH-DATE"))',
+                          '((:yyyymmdd 19440201))')
+
+        # TODO: fix or marked as expected to fail
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: These tests are commented out (lines 199-205). Make them into a separate method and mark them as expected failures:

```python
def test_dates_not_in_parens(self):
    [...]
```

File: tests/test_resolvers.py [hunk 13]
```diff
@@ -97,6 +289,165 @@ def test_first_paren(self):
         self.assertEqual(first_paren(txt), "dr. hello 2000-2012")
         self.assertIs(first_paren(txt_none), None)
 
+
+class TestInfoboxResolvers(TestResolvers):
+
+    def test_Jesus_birth_date_from_infobox_no_weird_behavior(self):
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: method name should be lowercase

File: tests/test_resolvers.py [hunk 14]
```diff
@@ -97,6 +289,165 @@ def test_first_paren(self):
...
+                                 msg=r'Parse {{plainlist|...}} template correctly in GWTW DIRECTOR')
+        self.assertRegexpMatches('(get "wikipedia-weapon" "M1 Abrams" (:code "WARS"))',
+                                 r'Gulf War',
+                                 msg='Returns link text, not link target')
+        self.assertRegexpMatches('(get-attributes "wikipedia-film" "Gone with the Wind (film)")',
+                                 r':code \"DIRECTOR\"')
+
+    def test_get_attributes_wikipedia_company(self):
+        self.assertRegexpMatches('(get-attributes "wikipedia-company" "BBC News")',
+                                 r':code \"KEY-PEOPLE\"')
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: no need to escape double quotes

File: tests/test_resolvers.py [hunk 15]
```diff
@@ -115,8 +466,8 @@ def test_image_data_logo_and_cover(self):
         imgs_logo2 = self.term_resolver.image_data("Shahid Afridi Foundation", "image-data")
         imgs_cover = self.term_resolver.image_data("Thoughts of Love", "image-data")
         self.assertIn("WOIO_CBS_19.png", imgs_logo[0]["file"])
-        self.assertEquals("SAF_Logo.png", imgs_logo2[0]["file"])
-        self.assertEquals("TOL%28Custom%29.jpg", imgs_cover[0]["file"])
+        self.assertEqual("SAF_Logo.png", imgs_logo2[0]["file"])
+        self.assertEqual("TOL%28Custom%29.jpg", imgs_cover[0]["file"])
```
Comment (alvaromorales) [inline tests/test_resolvers.py]: You're right in changing this, we should use `assertEqual` in preparation for Python 3.x: http://stackoverflow.com/questions/930995/assertequals-vs-assertequal-in-python

However, we should _only_ be using assertEqual. Please replace all mentions of assertEquals with assertEqual in all test files.

**Comment by alvaromorales** [comment]
Looking good. I left a few inline comments. Additionally:

* You need to remove examples.py.

    ```bash
    $ git rm tests/examples.py
    ```
* Why are image data tests under `TestErrorResolvers`? Create a separate test suite (class) `TestImageData` or put them in `TestTermResolvers`

**Comment by alvaromorales** [comment]
I don't understand why some lines have been changed from `self.assertEqual` to `unittest.TestCase.assertEqual`

**Comment by alvaromorales** [comment]
Ah, that makes sense. After fixing the merge conflicts, this looks good to go! :rocket:

**Comment by alvaromorales** [comment]
looks good! :ship:
